orca_detector/generator.py

```diff
@@ -1,94 +1,180 @@
-
-        return X,labels_temp
+        # Generate indexes of the batch
+        batch_indices = self.indices[index * self.batch_size: \
+                                     (index + 1) * self.batch_size]
+
+        # Store labels in a numpy array (batch_size, num_classes)
+        batch_labels = np.zeros((len(batch_indices), self.num_classes))
+        for batch_i, i in enumerate(batch_indices):
+            batch_labels[batch_i] = self.labels[i]
+
+        # Find list of IDs and pass in for feature extraction
+        batch_files  = [self.files[k] for k in batch_indices]
+        X = self.__extract_features(batch_files)
+        
+        # Deal with missing data in the X array
+        X, batch_labels = self.__treat_missing_data(X, batch_labels)
+        return X, batch_labels
 
     def on_epoch_end(self):
-        'Updates indexes after each epoch'
-        self.indexes = np.arange(len(self.files))
+        """Updates indices after each epoch"""
+
+        self.indices = np.arange(len(self.files))
         if self.shuffle == True:
-            np.random.shuffle(self.indexes)
-
-    def __extract_features(self, files_temp):
-        # X : (n_samples, *(frames_size, bands_size))
-        'Generates data containing batch_size samples'
-        # Initialization
-        X = []
-        # Generate data
-        for i, file in enumerate(files_temp):
-            data, sr = sf.read(
-                file.split(':')[0], 
-                start = int(file.split(':')[1]), 
-                frames = int(file.split(':')[2]))
-            # Store sample
-            X.append(_waveform_to_examples(data, sr))
+            np.random.shuffle(self.indices)
+
+    def __extract_features(self, batch_files):
+        """
+            Generates data containing samples for a given batch.  Differing 
+            from the original Keras VGGish implementation, each item in the
+            batch_files list here represents a segment of a wave file, so
+            they don't need to be further sub-divided.
+            
+            Return format is based on pretrained Keras VGGish model input shape.
+            
+            Returns X : np.array (num_samples, num_frames, num_bands, 1)
+            
+            IMPORTANT: if any samples were too small to be processed, there
+            will be rows in the returned X array which are filled with zeros.
+        """
+        
+        num_files = len(batch_files)
+        X = np.zeros((num_files, params.NUM_FRAMES, params.NUM_BANDS, 1))
+        
+        # Generate data from the appropriate segment of the audio file
+        for i, file in enumerate(batch_files):
+            data, sample_rate = sf.read(file.split(':')[0], 
+                                        start = int(file.split(':')[1]), 
+                                        frames = int(file.split(':')[2]))
+            # Transform to log mel spectrogram format and store sample
+            cur_spectro = _waveform_to_examples(data, sample_rate)
```

Comment:
Nit: Rename the cur_spectro  to something more meaningful eg. features

Reply:
I'll expand it to current_spectrogram (since it's inside a loop), but to me in terms of understanding this block of code, it's more helpful to be specific (i.e. spectrogram) than generic (i.e. features).